astropy/io/fits/header.py

```diff
@@ -350,13 +350,29 @@ def fromstring(cls, data, sep=''):
 
         cards = []
 
+        if not isinstance(sep, str):
```

Comment:
If `sep` and `data` are in bytes, `sep` is changed to `str` here, and then back to `bytes` below; I suggest to move this `if` stanza into the `else` for `data` that is `str` below.

Reply:
I agonized over this bc I agree it's really ugly.  My idea here was that 'sep' could be passed as either str or bytes and it will convert it to the right type for the right context.  I'll take a look at your suggestion.